### DISCUSSION TITLE
Machine-readable formal description of the PX format, such as EBNF

### DISCUSSION THREAD

**Discussion by mikaelhg** [statisticssweden/PxWeb#293] [Q&A] [answered]
Hello.

It's easy to find the SCB.se PX 2013 format PDF, and the accompanying website, but having handwritten a bunch of parsers and protocol serializers before, I'm not really relishing the idea of trying to make sure that I've really correctly understood everything in those documents.

Because of the well-known issues that crop up with this kind of a manual process, many communities that need to deal with a particular file format a lot, oftentimes create a formal machine-readable description of how to parse and validate data and files in those formats.

Does the PX community have something like this available in some repository somewhere? A EBNF description of the file format, an ANTLR G4 lexer and parser description, or something similar?

The ultimate aim I'm trying to accomplish here, is to bring the statistics offered as PX files, and through PxWeb in other formats, more readily available to an audience of data enthusiasts, whether scientists, ML folk or open sourcers. In order to accomplish this objective, I wish to offer the statistics data contained in PX files, in Parquet and Arrow formats, and develop a streaming converter which accomplishes this transformation on Linux, Mac and Windows, x86_64 and ARM, using only a few hundred kilobytes of memory, even on PX files of hundreds of megabytes.

Edit - current WIP at https://github.com/mikaelhg/pxspec/blob/master/README.md

```ebnf
header row = keyword , "=" , values , ";" , EOL ;
```

```ebnf
keyword = basekey , [ language ] , [ subkeys ] ;

basekey = "A".."Z" , { "A".."Z" | "0".."9" | "-" } ;

language = "[" , 2 * ( "a".."z" ) , "]" ;

subkeys = "(" , quoted-string-list , ")" ;

```

You have to order choices in a certain way for a left-recursive parser,
the order here is only a implementation detail.

```ebnf
values =
        integer
       | tlist-value
       | bare-string
       | hierarchy-levels
       | multiline-quoted-string-list
       ;

bare-string = { all characters - ( ";" | '"' ) }- ; (* one or more character *)

multiline-quoted-string =
        ( quoted-string , EOL , quoted-string , { EOL , quoted-string } )
       | quoted-string
       ;

quoted-string-list = quoted-string , { "," , quoted-string } ;

multiline-quoted-string-list =
        multiline-quoted-string , { "," , [ EOL ] , multiline-quoted-string ;
```

Specials:

```ebnf
tlist-value =
        "TLIST(" , time-scale , [ quoted-string , "-" , quoted-string ] , ")"
          , [ "," , multiline-quoted-string-list ] ;

time-scale = "A1" | "H1" | "Q1" | "M1" | "W1" ;

hierarchy-levels =
        quoted-string , "," , quoted-string , ":" , quoted-string ,
                      { "," , [ EOL ] , quoted-string , ":" , quoted-string } ;
```

Quoted strings in this format have no escaping whatsoever, you simply cannot express
the quote character as content in a PX string.

```ebnf
quoted-string = '"' , { all characters - '"' } , '"' ;

all characters = ? all visible characters ? ;

integer = [ "-" ] , "0".."9" , { "0".."9" } ;

EOL = "\r\n" | "\n" ;
```

**Comment by likp** [comment, marked as answer]
Hi @mikaelhg, it sound like an interesting project that you work on. But unfortunately we do not have any formal specification of the PX file format that you would be able to use to generate code from. 

I see how such a specification you be of great benefit although it would not be something that we would prioritize at the moment.

**Reply by mikaelhg** [reply]
Thanks. Has the Px community considered whether the future roadmap could include the adoption of an externally maintained file format suitable for sparse data cubes, such as Apache ORC?

**Reply by trygu** [reply]
I know that Statistics Ireland tried to get an initiative going, trying to formalize the specification of the PX-format, although it seems it's abandoned now (https://github.com/CSOIreland/PXS-Standard)

I think it would be hard to do without changing the format quite a lot. The format have evolved over the years, with its quirks and deviations. As @likp said; It would be a great benefit, but it would require a formalization of the format, which again would require a lot of code to be rewritten to support the formalized format. A new format based on Apache ORC or Parquet would be very useful, especially for large (or sparse) data cubes.

**Reply by mikaelhg** [reply]
One way to approach this would be to formalize the format through transcribing the current VB.net parser state machine from PCAxis.Core into a declarative, machine readable, format.

Is that PX parser the only canonical and complete one available as open source?

**Reply by likp** [reply]
> Thanks. Has the Px community considered whether the future roadmap could include the adoption of an externally maintained file format suitable for sparse data cubes, such as Apache ORC?

We currently support both PX files ans relational databases (Oracle and MS SQL Server) having the Common Nordic Meta Model, CNMM and in our ongoing work on the PxAPI 2.0 we are making it easier to add new types of data sources. Although we have no direct plan to implement such a data source our self's.

**Reply by likp** [reply]
> One way to approach this would be to formalize the format through transcribing the current VB.net parser state machine from PCAxis.Core into a declarative, machine readable, format.
> 
> Is that PX parser the only canonical and complete one available as open source?

As far as I know this is the only parser available as open source. PxEdit written by Statistics Finland should have it's own although I believe that the code is not release as open source.

**Comment by mikaelhg** [comment]
I ended up with three repositories for the specification and implementation of this tool:

* https://github.com/mikaelhg/pxspec - for the specification and Python prototype

* https://github.com/mikaelhg/px2csv-go - for the golang implementation, which takes about 1.2 seconds to parse and convert a 600 MB PX data cube file into CSV

* https://github.com/mikaelhg/px2csv-java - for the Java implementation, which takes about 2.0 seconds for the same job
